core/src/letter_distribution.c

```diff
@@ -153,47 +153,51 @@ uint8_t human_readable_letter_to_machine_letter(
 // Convert a string of arbitrary characters into an array of machine letters,
 // returning the number of machine letters. This function does not allocate
 // the ml array; it is the caller's responsibility to make this array big
-// enough.
+// enough. This function will return -1 if it encounters an invalid letter.
 // Note: This is a slow function that should not be used in any hot loops.
 int str_to_machine_letters(const LetterDistribution *letter_distribution,
                            const char *str, bool allow_played_through_marker,
                            uint8_t *mls) {
 
   int num_mls = 0;
   int num_bytes = string_length(str);
-  int i = 0;
-  int prev_i = -1;
-  while (i < num_bytes) {
-    for (int j = i + letter_distribution->max_tile_length; j > i; j--) {
-      if (j > num_bytes) {
-        continue;
+  // Use +1 for the null terminator
+  char current_letter[MAX_LETTER_CHAR_LENGTH + 1];
+  int current_letter_byte_index = 0;
+  bool building_multichar_letter = false;
+
+  for (int i = 0; i < num_bytes; i++) {
+    char current_char = str[i];
```

Comment:
`for (char current_char; (current_char = *str); ++str) { ... }` may be faster than two passes (`string_length` is O(n))